### PR TITLE
Permutation function for numpy frontend Ivy API

### LINKED ISSUE #9451: permutation
Add [permutation](https://github.com/unifyai/ivy/tree/master/ivy/functional/frontends/numpy/random) to function frontend and respective unitest.

ToDo issue:  #1530

### PR DESCRIPTION
**Pull request by Tiago-B-C-Reis**
Close #9451

### CODE REVIEW

**Comment by Tiago-B-C-Reis** [comment]
Hello @ivy-leaves,

I'm creating this PR to add permutation function.

**Comment by Tiago-B-C-Reis** [comment]
Hello [ukj02](https://github.com/ukj02),

I have submitted my PR but the unitest is returning these errors:
`IndexError: tuple index out of range`
`TypeError: len() of unsized object`
`TypeError: Scalar tensor has no len()`
`TypeError: len() of unsized object`
`TypeError: len() of a 0-d tensor`

They all seem to be related to "len()" function, but I'm not seeing why, could you give me some hint/help on this, please?

**Comment by ukj02** [comment]
Hi @Tiago-B-C-Reis , thank you for the contribution!
While looking at the errors I see that you get an `IndexError: tuple index out of range`. This is because you are trying to access an item in a tuple that does not exist. I think you are using the `len()` function wrong and have to reimplement the frontend code.

Thanks! 😄

**Comment by ukj02** [comment]
Close #9451

**Comment by Tiago-B-C-Reis** [comment]
Hello @ukj02,
Thank you very much for the feedback 😄

Just have one question, since you have close #9451 that means that this PR can be closed, or firstly I need to correct that error?

**Comment by ukj02** [comment]
> Just have one question, since you have close #9451 that means that this PR can be closed, or firstly I need to correct that error?
You do not need to now but for future PRs you should remember to add the issue the PR is closing.

**Comment by Tiago-B-C-Reis** [comment]
Hello @ukj02 ,

I have reviewed the code on the frontend, and now it seems that I only have this error: (ignore the formatting please)
image

I'm not seeing what is causing it, can you please help here? or if you detect other possible problems, please say.
Thank you for the support until now :)

**Comment by ukj02** [comment]
The error's I see on my end are just formatting or the ones I have mentioned above. Nothing else has changed. Once you think you have a solution push it so I can see the changes.

**Comment by Tiago-B-C-Reis** [comment]
Hello [ukj02](https://github.com/ukj02),

I'm having difficulties doing the Ivy front-end test. 
I have already followed all the instructions on the [Ivy Frontend Tests](https://lets-unify.ai/ivy/deep_dive/ivy_frontends_tests.html) and read some examples, but still, I'm not getting it right. 
Could you please help here?

Thank you a lot.

**Comment by guillesanbri** [comment]
Hi! I'll be reviewing this PR now :slightly_smiling_face:

File: ivy/functional/frontends/numpy/random/functions.py [hunk 1]
```diff
@@ -39,3 +41,39 @@ def geometric(p, size=None):
...
 def normal(loc=0.0, scale=1.0, size=None):
     return ivy.random_normal(mean=loc, std=scale, shape=size, dtype="float64")
+
+
+@to_ivy_arrays_and_back
+@from_zero_dim_arrays_to_scalar
+def permutation(x):
+    if isinstance(x, int):
+        new_array = ivy.arange(x)
+        return new_array
```
Comment (guillesanbri) [inline ivy/functional/frontends/numpy/random/functions.py:51]: I think this wouldn't get permuted :slightly_smiling_face:

File: ivy/functional/frontends/numpy/random/functions.py [hunk 2]
```diff
@@ -39,3 +41,39 @@ def geometric(p, size=None):
...
+            position_list = []
+            arr_list = []
+            while len(position_list) < first_dim:
+                r_position = round(rd.uniform(1.0, first_dim))
+                if r_position not in position_list:
+                    position_list.append(r_position)
+                    arr_list.append(x[r_position - 1])
+
+            new_array = ivy.array(arr_list).reshape(arr_shape)
+            return new_array
```
Comment (guillesanbri) [inline ivy/functional/frontends/numpy/random/functions.py:76]: Probably all of this can be replaced with [`ivy.shuffle`](https://lets-unify.ai/ivy/functional/ivy/random.html#ivy.functional.ivy.random.shuffle)

File: ivy_tests/test_ivy/test_frontends/test_numpy/test_random/test_functions.py [hunk 3]
```diff
@@ -136,36 +136,30 @@ def test_numpy_normal(
...
-    input_dtypes=helpers.get_dtypes("float"),
-    p=st.floats(
-        allow_nan=False,
-        allow_infinity=False,
-        width=32,
-        min_value=9.999999747378752e-06,
-        max_value=0.9999899864196777,
-    ),
-    size=st.tuples(
-        st.integers(min_value=2, max_value=5), st.integers(min_value=2, max_value=5)
```
Comment (guillesanbri) [inline ivy_tests/test_ivy/test_frontends/test_numpy/test_random/test_functions.py:150]: seems like you have overwritten the `geometric` test, you should instead write a new test for the new function :slightly_smiling_face:

**Review by guillesanbri — CHANGES_REQUESTED** [review]
just requested a few changes, thank you!

**Comment by guillesanbri** [comment]
Hello again, just noticed that the issue has not been reserved in the ToDo list, please follow the steps on the contributing page (https://lets-unify.ai/ivy/contributing/the_basics.html#todo-list-issues) so that nobody else works on this function :slightly_smiling_face:

**Comment by Tiago-B-C-Reis** [comment]
Hello [guillesanbri](https://github.com/guillesanbri)

Thank you very much for all the feedback, you are the best :)

It seems that I correct those errors that you mention, but now I have a new one:
`fixture 'size' not found
`
Does this mean that I need to remove the "size" from the test?

File: ivy_tests/test_ivy/test_frontends/test_numpy/test_random/test_functions.py [hunk 4]
```diff
@@ -203,3 +203,31 @@ def test_numpy_geometric(
...
+    input_dtype, x = dtype_and_x
+    helpers.test_frontend_function(
+        input_dtypes=input_dtype,
+        frontend=frontend,
+        test_flags=test_flags,
+        fn_tree=fn_tree,
+        on_device=on_device,
+        x=x[0],
+        test_values=False,
+        size=size,
```
Comment (guillesanbri) [inline ivy_tests/test_ivy/test_frontends/test_numpy/test_random/test_functions.py:232]: indeed, as the tested function doesn't take any `size` argument, you should remove it from here and also from above :slightly_smiling_face: also probably good to move the `x=x[0]` to be the last argument (this won't make any difference but just general styling) :grin:

**Review by guillesanbri — COMMENTED** [review]
Just left a comment about the `size` arg :slightly_smiling_face:

**Comment by Tiago-B-C-Reis** [comment]
Hello again [guillesanbri](https://github.com/guillesanbri)
Thank you again for your kind support.

It seems that the test is running ok now, but I'm having some errors related to the function, one I think is related to this part of the function:
```
elif isinstance(x, str):
        return "IndexError: x must be an integer " \
               "or at least 1-dimensional"
```
I also have this error on line 78 of the function:
`IndexError: tuple index out of range`

Honestly, I don't know what is causing these errors, do you have any hint for me?
Nevertheless, I going to work on it.

File: ivy/functional/frontends/numpy/random/functions.py [hunk 5]
```diff
@@ -61,3 +61,19 @@ def multinomial(n, pvals, size=None):
...
+@to_ivy_arrays_and_back
+@from_zero_dim_arrays_to_scalar
+def permutation(x):
+    if isinstance(x, int):
+        new_array = ivy.arange(x)
+        ivy.random.shuffle(new_array)
+        return new_array
+    elif isinstance(x, str):
+        return "IndexError: x must be an integer " \
+               "or at least 1-dimensional"
```
Comment (guillesanbri) [inline ivy/functional/frontends/numpy/random/functions.py:75]: not sure why you are checking for `str` here :thinking:

File: ivy_tests/test_ivy/test_frontends/test_numpy/test_random/test_functions.py [hunk 6]
```diff
@@ -235,3 +235,29 @@ def test_numpy_multinomial(
         pvals=np.array([1 / n] * n, dtype=dtype[0]),
         size=size,
     )
+
+
+# permutation
+@handle_frontend_test(
+    fn_tree="numpy.random.permutation",
+    dtype_and_x=helpers.dtype_and_values(
+        available_dtypes=helpers.get_dtypes("numeric"),
```
Comment (guillesanbri) [inline ivy_tests/test_ivy/test_frontends/test_numpy/test_random/test_functions.py:244]: probably need to add `min_num_dims=1` here so that arrays with zero dimensions are not generated :slightly_smiling_face:

**Review by guillesanbri — CHANGES_REQUESTED** [review]
just left a couple of comments more, the one regarding the number of dimensions should probably help with your latest issue!

**Comment by Tiago-B-C-Reis** [comment]
Hello [guillesanbri](https://github.com/guillesanbri)

I have improved the function and ivy test and now I'm having only this error:

```
ivy/functional/frontends/numpy/random/functions.py:80: in permutation
  raise ValueError("x must be an integer or at least 1-dimensional")
  ValueError: x must be an integer or at least 1-dimensional
  Falsifying example: test_numpy_permutation(
    dtype_and_x=(['uint8'], [array(0, dtype=uint8)]),
    fn_tree='ivy.functional.frontends.numpy.random.permutation',
    test_flags=num_positional_args=0. with_out=False. inplace=False. native_arrays=[False]. as_variable=[False]. ,
    frontend='numpy',
    on_device='cpu',
)
```

I'm not applying your hint to add this `min_num_dims=1` correctly?

File: ivy/functional/frontends/numpy/random/functions.py [hunk 7]
```diff
@@ -61,3 +61,20 @@ def multinomial(n, pvals, size=None):
...
+        ivy.random.shuffle(new_array)
+        return new_array
+    elif isinstance(x, str):
+        raise ValueError("x must be an integer or at least 1-dimensional")
+    elif x.ndim > 0:
+        new_array = x
+        ivy.random.shuffle(new_array)
+        return new_array
+    else:
+        raise ValueError("x must be an integer or at least 1-dimensional")
```
Comment (guillesanbri) [inline ivy/functional/frontends/numpy/random/functions.py:80]: a couple of things:
1.- There is no real need to do all the checks here, the backends will raise these errors.
2.- `ivy.shuffle` doesn't operate in-place, so you are not actually returning the shuffled array :slightly_smiling_face: 

probably this can all be summed up in 
```python
def permutation(x):
    if isinstance(x, int):
        x = ivy.arange(x)
    return ivy.shuffle(x)
```

File: ivy_tests/test_ivy/test_frontends/test_numpy/test_random/test_functions.py [hunk 8]
```diff
@@ -235,3 +235,30 @@ def test_numpy_multinomial(
...
+        min_num_dims=1
+    ),
+)
+def test_numpy_permutation(
+    dtype_and_x,
+    frontend,
+    test_flags,
```
Comment (guillesanbri) [inline ivy_tests/test_ivy/test_frontends/test_numpy/test_random/test_functions.py:251]: I can replicate the failure that you are facing, however, I do get an error which I believe it's on our side. I'll ask about it and update soon!

**Review by guillesanbri — CHANGES_REQUESTED** [review]
Left a couple of comments, let me know if I can help with anything :slightly_smiling_face:

**Comment by guillesanbri** [comment]
Regarding the last error, changing the signature of `permutation` to `def permutation(x, /):` fixes it :slightly_smiling_face:

File: ivy/functional/frontends/numpy/random/functions.py [hunk 9]
```diff
@@ -61,3 +61,13 @@ def multinomial(n, pvals, size=None):
...
     return ivy.multinomial(n, num_samples, batch_size=batch_size, probs=pvals)
+
+
+@to_ivy_arrays_and_back
+@from_zero_dim_arrays_to_scalar
+def permutation(x, /):
+    if isinstance(x, int):
+        x = ivy.arange(x)
+    elif x.ndim > 0:
+        x = x
```
Comment (guillesanbri) [inline ivy/functional/frontends/numpy/random/functions.py:72]: we can remove these two lines, as the output is the same whether `x.ndim > 0` or not

**Review by guillesanbri — CHANGES_REQUESTED** [review]

**Comment by Tiago-B-C-Reis** [comment]
Hello [guillesanbri](https://github.com/guillesanbri) :)

In the Numpy documentation for this permutation function, it says that has 2 conditions, one is if the x input is an integer and the other if the x input is an array. 
image

If I do like this, does the function only works when the x input is an integer?
```
def permutation(x):
    if isinstance(x, int):
        x = ivy.arange(x)
    return ivy.shuffle(x)
```

Another doubt, regarding the way ivy.shufle works. This is an example on the documentation:
```
 >>> x = ivy.array([1, 2, 3, 4, 5])
    >>> y = ivy.shuffle(x)
    >>> print(y)
    ivy.array([2, 1, 4, 3, 5])
```

Following this example, should it be like this?
```
def permutation(x, /):
    if isinstance(x, int):
        x_array = ivy.arange(x)
        new_array = ivy.shuffle(x_array)
        return new_array
    elif x.ndim > 0:
        new_array = x
        new_array = ivy.shuffle(new_array)
        return new_array
```

In summary, I'm not right if I'm applying the Ivy.shuffle correctly since it's different from np.shuffle and also the first question I describe above.

I hope that I was able to expose my doubts clearly and tI hope you can help as you have been done already :)

**Comment by guillesanbri** [comment]
Hi, regarding your questions

> If I do like this, does the function only works when the x input is an integer?

Not really, if the input is an integer then `x` is converted into an array and then you shuffle that array, if `x` is not an integer, (i.e. an array) it will get passed directly to `ivy.shuffle`, which will raise the corresponding error if we are passing another data type (e.g. a string)

 > Following this example, should it be like this?

I don't fully follow what's the problem here, if you look at the code, it's doing the same but with extra variables. The only difference would be the `ndim` check which will also throw an error if the input is not an array so it's not doing much there. `ivy.shuffle` does not operate inplace, so the copy of the array is happening inside the function (btw, doing something like `x = x` which you had in a previous commit does not copy the array, it simply references it with a different name) :slightly_smiling_face:

**Comment by guillesanbri** [comment]
other than that, the PR seems ready to go, so I'll be merging this one, let me know if there is anything that I can help with!

**Comment by Tiago-B-C-Reis** [comment]
Hi [guillesanbri](https://github.com/guillesanbri)

I just want to thank you for taking my doubts and helping me with this process 🙂
Keep up the great job!

**Comment by guillesanbri** [comment]
thank you for the contribution! :grin:
